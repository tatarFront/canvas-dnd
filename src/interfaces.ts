export interface ICanvasElement {
  color: string;
  canvasPoints: number[];
}

export interface ICanvas extends Object {
  firstCanvas: Array<ICanvasElement>;
  secondCanvas: Array<ICanvasElement>;
  error: string;
}

export interface IAction {
  payload: any;
  type: string;
}

export interface IState {
  canvas: ICanvas;
}
