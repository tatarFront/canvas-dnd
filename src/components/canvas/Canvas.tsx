//NOTE: This is a component-section for lines
import React, { useRef, useEffect, useCallback, MouseEvent } from 'react';

import { ICanvasElement, ICanvas } from '../../interfaces';
import { canvasSettings } from '../../tools';
interface IProps {
  data?: ICanvasElement[];
  keyName: keyof ICanvas;

  canvasWrapRef?: any;
  setDrag?: any;

  create: (keyName: keyof ICanvas) => void;
  replace: (
    keyName: keyof ICanvas,
    points: number[],
    nextPoints: number[]
  ) => void;
  update: (
    keyName: keyof ICanvas,
    prevPoints: number[],
    nextPoints: number[]
  ) => void;
}

const Canvas: React.FC<IProps> = ({
  data,
  keyName,
  create,
  replace,
  update,
  canvasWrapRef,
  setDrag
}) => {
  const canvasRef: any = useRef(null);
  const prevDataRef: any = useRef([]);
  const dragLine: any = useRef(null);

  useEffect(() => init());

  const init = () => {
    if (canvasRef.current && !compareData(data, prevDataRef.current)) {
      const ctx = canvasRef.current ? canvasRef.current.getContext('2d') : null;
      if(ctx) {
        ctx.clearRect(0, 0, canvasSettings.width, canvasSettings.heigth);
        data &&
          data.forEach(item => {
            createLine(item.canvasPoints, item.color);
          });
  
        ctx.beginPath();
        ctx.stroke();
      }

      //NOTE: save prevProps
      prevDataRef.current = data;
    }
  };

  const generate2PathElement = (points: number[]) => {
    const [a1, b1, a2, b2, a3, b3] = points;
    const line = new Path2D();
    line.moveTo(a1, b1);
    line.lineTo(a2, b2);
    line.lineTo(a3, b3);

    return line;
  };

  const createLine = (points: any, color: string) => {
    const ctx = canvasRef.current.getContext('2d');
    const line = generate2PathElement(points);

    ctx.lineWidth = canvasSettings.lineWidth;
    ctx.beginPath();
    ctx.strokeStyle = color;
    ctx.stroke(line);
  };

  const compareData = useCallback((currentData, nextData) => {
    if (currentData.length !== nextData.length) return false;

    return !currentData.some((curItem: ICanvasElement, i: number) => {
      const nextItem = nextData[i];
      return JSON.stringify(curItem) !== JSON.stringify(nextItem);
    });
  }, []);

  const onMouseDown = function(e: MouseEvent) {
    const canvas = canvasRef.current;
    const ctx = canvas.getContext('2d');
    const mousePos = {
      x: e.pageX - canvas.offsetLeft,
      y: e.pageY - canvas.offsetTop
    };

    data &&
      data.some((item: any): any => {
        const line = generate2PathElement(item.canvasPoints);
        if (ctx.isPointInStroke(line, mousePos.x, mousePos.y)) {
          setDrag(true);

          dragLine.current = { ...item, canvasPointsPrev: item.canvasPoints };
          const canvasWrap = canvasWrapRef.current;

          if (canvasWrap && dragLine.current) {
            const [a1, b1, a2, b2, a3, b3] = item.canvasPoints;
            const minA = Math.min(a1, a2, a3) - canvasSettings.lineWidth / 2;
            const minB = Math.min(b1, b2, b3) - canvasSettings.lineWidth / 2;

            const lineCopy = generate2PathElement([
              a1 - minA,
              b1 - minB,
              a2 - minA,
              b2 - minB,
              a3 - minA,
              b3 - minB
            ]);
            const ctxWrap = canvasWrap.getContext('2d');
            ctxWrap.lineWidth = canvasSettings.lineWidth;
            ctxWrap.strokeStyle = dragLine.current.color;
            ctxWrap.beginPath();
            ctxWrap.stroke(lineCopy);
          }

          canvas.onmousemove = onMouseMove;
          canvasWrapRef.current.onmousemove = onMouseMove;
          canvasWrapRef.current.onmouseup = onMouseUp;
          return true;
        }

        return false;
      });
  };

  function onMouseMove(e: MouseEvent) {
    if (dragLine.current) {
      const [a1, b1, a2, b2, a3, b3] = dragLine.current.canvasPoints;
      dragLine.current.canvasPoints = [
        a1 + e.movementX,
        b1 + e.movementY,
        a2 + e.movementX,
        b2 + e.movementY,
        a3 + e.movementX,
        b3 + e.movementY
      ];
    }
  }

  function onMouseUp(e: MouseEvent) {
    setDrag(false);
    const canvasWrap = canvasWrapRef.current;
    const canvas = canvasRef.current;
    const mousePos = {
      x: e.pageX - canvas.offsetLeft,
      y: e.pageY - canvas.offsetTop
    };
    if (dragLine.current) {
      const [a1, b1, a2, b2, a3, b3] = dragLine.current.canvasPoints;
      if (mousePos.y > canvasSettings.heigth || mousePos.y <= 0) {
        const coefficient =
          keyName === 'firstCanvas'
            ? -1 * canvasSettings.k
            : canvasSettings.k + 20;

        const nextPoints = [
          a1,
          b1 - mousePos.y - coefficient,
          a2,
          b2 - mousePos.y - coefficient,
          a3,
          b3 - mousePos.y - coefficient
        ];

        const isBeside =
          Math.max(nextPoints[1], nextPoints[3], nextPoints[5]) < 0 ||
          Math.min(nextPoints[1], nextPoints[3], nextPoints[5]) >
            canvasSettings.heigth;
        replace(
          keyName,
          dragLine.current.canvasPointsPrev,
          isBeside ? dragLine.current.canvasPointsPrev : nextPoints
        );
      } else {
        update(
          keyName,
          dragLine.current.canvasPointsPrev,
          dragLine.current.canvasPoints
        );
      }
      dragLine.current = null;
    }

    if (canvasWrap) {
      const ctxWrap = canvasWrapRef.current
        ? canvasWrapRef.current.getContext('2d')
        : null;
      ctxWrap.clearRect(0, 0, canvasSettings.width, canvasSettings.heigth);
      canvas.onmousemove = null;
      canvasWrapRef.current.onmousemove = null;
      canvasWrapRef.current.onmouseup = null;
    }
  }

  return (
    <div
      className="shadow p-3 mb-5 bg-white rounded"
      style={{ backgroundColor: 'white' }}
    >
      <button
        type="button"
        className="btn btn-outline-primary"
        onClick={() => create(keyName)}
      >
        Create new line
      </button>
      <div>
        <canvas
          id={keyName}
          ref={canvasRef}
          width={canvasSettings.width}
          height={canvasSettings.heigth}
          onMouseDown={onMouseDown}
          onMouseUp={onMouseUp}
        />
      </div>
    </div>
  );
};

export default Canvas;
