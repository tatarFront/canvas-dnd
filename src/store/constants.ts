export default {
  CREATE_LINE_SUCCESS: 'CREATE_LINE_SUCCESS',
  REPLACE_LINE_SUCCESS: 'REPLACE_LINE_SUCCESS',
  UPDATE_LINE_SUCCESS: 'UPDATE_LINE_SUCCESS',
};
