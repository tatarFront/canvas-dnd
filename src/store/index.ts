import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducer';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}
const composeEnhancers =
  (typeof window !== 'undefined' &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;

export default createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk))
);
