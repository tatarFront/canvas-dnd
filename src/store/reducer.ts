import { combineReducers } from 'redux';

import { ICanvas, IAction, IState } from '../interfaces';
import types from './constants';

export const initialState: ICanvas = {
  firstCanvas: [],
  secondCanvas: [],
  error: ''
};

export const canvasState = (state: ICanvas = initialState, action: IAction) => {
  switch (action.type) {
    case types.CREATE_LINE_SUCCESS:
    case types.UPDATE_LINE_SUCCESS:
    case types.REPLACE_LINE_SUCCESS:
      return {
        ...state,
        ...action.payload
      };

    default:
      return state;
  }
};

export default combineReducers<IState>({
  canvas: canvasState
});
