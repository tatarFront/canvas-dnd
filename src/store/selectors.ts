import { createSelector } from 'reselect';
import { IState, ICanvas } from '../interfaces';

export const getCanvas = (state: IState) => state.canvas;

export const getData = (state: IState, keyName: keyof ICanvas) =>
  createSelector(getCanvas, canvas => canvas[keyName])(state);

export default {
  getCanvas,
  getData
};
