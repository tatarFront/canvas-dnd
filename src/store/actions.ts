import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { ICanvasElement, IAction, ICanvas } from '../interfaces';
import { getCanvas, getData } from './selectors';
import types from './constants';
import { compareArrays, randomInteger, getRandomColor } from '../tools';

//NOTE: actions
export const createLineSuccess = (payload: object) => ({
  type: types.CREATE_LINE_SUCCESS,
  payload
});

export const replaceLineSuccess = (payload: object) => ({
  type: types.REPLACE_LINE_SUCCESS,
  payload
});

export const updateLineSuccess = (payload: object) => ({
  type: types.UPDATE_LINE_SUCCESS,
  payload
});

//NOTE: middleware Thunk
export const create = (
  keyName: keyof ICanvas
): ThunkAction<any, any, any, any> => {
  return (dispatch: ThunkDispatch<{}, {}, IAction>, getState: any) => {
    const state = getState();
    const newCanvasElement: ICanvasElement = {
      color: getRandomColor(),
      canvasPoints: [
        randomInteger(0, 100),
        randomInteger(0, 180),

        randomInteger(110, 200),
        randomInteger(0, 180),

        randomInteger(210, 300),
        randomInteger(0, 180)
      ]
    };
    dispatch(
      createLineSuccess({
        [keyName]: [...state.canvas[keyName], newCanvasElement]
      })
    );
  };
};

const generateNextData = (
  arr1: ICanvasElement[],
  arr2: ICanvasElement[],
  nextPoint: number[],
  cb: (points: number[]) => void | boolean
) => {
  const [newItem] = arr1.filter((item: ICanvasElement) =>
    cb(item.canvasPoints)
  );
  const second: Array<ICanvasElement> = [
    ...arr2,
    {
      ...newItem,
      canvasPoints: nextPoint
    }
  ];
  const first: Array<ICanvasElement> = arr1.filter(
    (item: any) => !cb(item.canvasPoints)
  );

  return { first, second };
};

export const replace = (
  keyName: keyof ICanvas,
  points: number[],
  nextPoint?: number[]
): ThunkAction<any, any, any, any> => {
  return (dispatch: ThunkDispatch<{}, {}, IAction>, getState: () => any) => {
    const canvas: any = getCanvas(getState());
    const [first, second] =
      keyName === 'firstCanvas'
        ? ['firstCanvas', 'secondCanvas']
        : ['secondCanvas', 'firstCanvas'];

    const nextData = generateNextData(
      canvas[first],
      canvas[second],
      nextPoint || points,
      (itemPoints: number[]) => compareArrays(itemPoints, points)
    );

    dispatch(
      replaceLineSuccess({
        [first]: nextData.first,
        [second]: nextData.second
      })
    );
  };
};

export const update = (
  keyName: keyof ICanvas,
  prevPoints: number[],
  nextPoints: number[]
): ThunkAction<any, any, any, any> => {
  return (dispatch: ThunkDispatch<{}, {}, IAction>, getState: any) => {
    const canvasData: any = getData(getState(), keyName);
    const nextCanvasData = canvasData.map((item: any) => {
      if (!compareArrays(item.canvasPoints, prevPoints)) return item;

      return {
        color: item.color,
        canvasPoints: nextPoints
      };
    });

    dispatch(
      updateLineSuccess({
        [keyName]: [...nextCanvasData]
      })
    );
  };
};
