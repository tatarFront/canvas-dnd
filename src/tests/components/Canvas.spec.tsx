import React from 'react';
import { Provider } from 'react-redux';

import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import store from '../../store';
import { ICanvas } from '../../interfaces';

import Canvas from '../../components/canvas/Canvas';
import CanvasContainer from '../../components/canvas';
import CanvasWrapper from '../../components/canvasWrapper';

const keyName: keyof ICanvas = 'firstCanvas';
const props = {
  data: [],
  keyName: keyName,
  create: () => {},
  replace: () => {},
  update: () => {}
};

describe('Canvas', () => {
  it('should render correctly container in "debug" mode', () => {
    const component = shallow(
      <Provider store={store}>
        <CanvasContainer keyName={keyName} />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
  
  it('should click and render correctly', () => {
    const component = mount(
      <Provider store={store}>
        <CanvasContainer keyName={keyName} />
      </Provider>
    );

    const canvasElement = component.find("button")
    canvasElement.simulate("click")
    expect(component).toMatchSnapshot();
  });

  it('should render correctly component in "debug" mode', () => {
    const component = shallow(<Canvas {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('render correct component prop type', () => {
    const CanvasComponent = mount(<Canvas {...props} />);

    expect(typeof CanvasComponent.props().create).toBe('function');
    expect(typeof CanvasComponent.props().update).toBe('function');
    expect(typeof CanvasComponent.props().replace).toBe('function');
    expect(typeof CanvasComponent.props().keyName).toBe('string');
  });

  it('render correct component for check elements', () => {
    const CanvasComponent = mount(<Canvas {...props} />);
    const canvasElement = CanvasComponent.find("canvas")
    expect(canvasElement).toHaveLength(1);
  });
});

describe('Canvas Wrapper', () => {
  it('should render correctly in "debug" mode', () => {
    const component = shallow(
      <CanvasWrapper>
        <CanvasContainer keyName="firstCanvas" />
        <CanvasContainer keyName="secondCanvas" />
      </CanvasWrapper>
    );
    expect(component).toMatchSnapshot();
  });
});

configure({ adapter: new Adapter() });
