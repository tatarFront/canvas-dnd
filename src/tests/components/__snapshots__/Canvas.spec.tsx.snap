// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas Wrapper should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`Canvas should click and render correctly 1`] = `ReactWrapper {}`;

exports[`Canvas should render correctly component in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`Canvas should render correctly container in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`Canvas should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;

exports[`Canvas should render correctly in "debug" mode 2`] = `ShallowWrapper {}`;
